libcob/profiling.c

```diff
@@ -0,0 +1,464 @@
+	   of all paragraphs of the section */
+	info->called_count[proc_idx] ++;
+}
+
+void
+cob_prof_enter_program (struct cob_prof_module *info){
+	/* program always has index 0 */
+	cob_prof_enter_procedure (info, 0);
+}
+
+void
+cob_prof_exit_program (struct cob_prof_module *info){
+	/* program always has index 0 */
+	cob_prof_exit_procedure (info, 0);
+}
+
+void
+cob_prof_use_paragraph_entry (struct cob_prof_module *info,
+			      int paragraph_idx, int entry_idx){
+	if (!is_active || info == NULL) return;
+	info->called_count[entry_idx] ++;
+	cob_prof_enter_paragraph (info, paragraph_idx);
+}
+
+void
+cob_prof_exit_paragraph (struct cob_prof_module *info, int proc_idx)
+{
+	cob_prof_exit_procedure (info, proc_idx);
+}
+
+void
+cob_prof_exit_section (struct cob_prof_module *info, int proc_idx)
+{
+	/* For now, nothing to do */
+}
+
+void
+cob_prof_goto (struct cob_prof_module *info)
+{
+	int curr_proc;
+	struct cob_prof_module *curr_info;
+
+	if (!is_active || info == NULL) return;
+
+	curr_proc = called_procedures[current_idx];
+	curr_info = called_runtimes[current_idx];
+
+	if (curr_info->procedures[curr_proc].kind == COB_PROF_PROCEDURE_PARAGRAPH){
+		cob_prof_exit_procedure (curr_info, curr_proc);
+	}
+}
+
+static void
+print_monotonic_time (FILE *file, cob_ns_time t) {
+
+	cob_ns_time nanoseconds = t ;
+	unsigned int hours = 0;
+	unsigned int minutes = 0;
+	unsigned int seconds = 0;
+	unsigned int milliseconds = 0;
+
+	if (nanoseconds >= 1000000) {
+		milliseconds = t / 1000000;
+		nanoseconds %= 1000000;
+	}
+	if (milliseconds >= 1000) {
+		seconds = milliseconds / 1000;
+		milliseconds %= 1000;
+	}
+	if (seconds >= 60) {
+		minutes = seconds / 60;
+		seconds %= 60;
+	}
+	if (minutes >= 60) {
+		hours = minutes / 60;
+		minutes %= 60;
+	}
+
+	fprintf (file, "%Ld,", t);
+
+	if (hours > 0) {
+		fprintf (file, "%u h %u m %u s", hours, minutes, seconds);
+	} else if (minutes > 0) {
+		fprintf (file, "%u m %u s %u ms", minutes, seconds, milliseconds);
+	} else if (seconds > 0) {
+		fprintf (file, "%u s %u ms %Lu ns", seconds, milliseconds, nanoseconds);
+	} else if (milliseconds > 0) {
+		fprintf (file, "%u ms %Lu ns", milliseconds, nanoseconds);
+	} else {
+		fprintf (file, "%Lu ns", nanoseconds);
+	}
+}
+
+void
+cob_prof_end ()
+{
+	FILE *file;
+	char prof_file_buf[COB_NORMAL_BUFF];
+	const char* prof_filename = NULL;
+	struct cob_prof_module_list *l;
+
+	prof_init_static ();
+
+	if (!cobsetptr || !is_active || !prof_info_list) return;
+
+	while (current_idx >= 0) {
+		cob_prof_exit_procedure (called_runtimes[current_idx],
+					 called_procedures[current_idx]);
+	}
+
+	prof_filename = cobsetptr->cob_prof_filename;
+	if (!prof_filename){
+		const char* program_id;
+
+		/* Do not use cob_sys_getpid() in case of fork() */
+		int pid = getpid();
+
+		for (l = prof_info_list ; l != NULL; l=l->next){
+			program_id = l->info->procedures[0].text;
+		}
+		sprintf(prof_file_buf, "cob-prof-%s-%d.csv", program_id, pid);
```

Comment:
Please put the `prof_filename` to `cob_expand_env_string()` instead.

Reply:
Ah, that's the very first profiled module, something `cob_expand_env_string` has no info about.
Is it really useful to use that instead of the main program?

`cob_expand_env_string()`  has already a note that something like $b/$d/$t should be added, that would be a good reason to add it and change the default for prof_filename in common.c to `cob-prof-$b-$$.csv`.

Reply:
I used $m (last program-id). $b would be `prog.exe` or `prog` ?

Reply:
> I used $m (last program-id). $b would be `prog.exe` or `prog` ?

this would be `prog` or `cobcrun`, no ".exe" there (.exe within a file name triggers some warnings somehow :-)

Reply:
That's more tricky, I will have to either use some static space or allocate or free afterwards (previous version would just return a pointer into the existing string)